File: trunk/core/src/main/groovy/org/tellurium/tool/Node.java
package org.tellurium.tool;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;

/**
 * A node for the tree
 *
 * @author Jian Fang (John.Jian.Fang@gmail.com)
 *
 *         Date: Dec 10, 2008
 */
public class Node {
    public static final String TAG = "tag";
    
    private String id;

    //relative xpath
    private String xpath;

    private Map<String, String> attributes = new HashMap<String, String>();

    private Node parent;

    private LinkedList<Node> children = new LinkedList<Node>();

    public void printUI(){
        boolean hasChildren = false;
        if(children.size() > 0)
            hasChildren = true;
        StringBuffer sb = new StringBuffer();
        sb.append(Ui.getType(attributes.get(TAG), hasChildren)).append("(UID: ").append(id).append(", clocator: [:])");
        if(hasChildren)
            sb.append("{");
        System.out.println(sb.toString());
        if(hasChildren){
            for(Node node: children){
                node.printUI();
            }
        }
        if(hasChildren)
            System.out.println("}");
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getXpath() {
        return xpath;
    }

    public void setXpath(String xpath) {
        this.xpath = xpath;
    }

    public Node getParent() {
        return parent;
    }

    public void setParent(Node parent) {
        this.parent = parent;
    }

    public LinkedList<Node> getChildren() {
        return children;
    }

    public void setChildren(LinkedList<Node> children) {
        this.children = children;
    }

    public Map getAttributes() {
        return attributes;
    }

    public void setAttributes(Map attributes) {
        this.attributes = attributes;
    }

    public boolean isEmpty(){
        if(children != null && children.size() > 0)
            return true;
        
        return false;
    }

    public void addChild(Node child){
        children.add(child);
    }
    
    public void removeChild(String uid){
        Node child = findChild(uid);
        if(child != null){
            children.remove(child);
        }
    }

    public Node findChild(String uid){
        for(Node current: children){
            if(current.getId().equals(uid)){
                return current;
            }
        }

        return null;
    }
}

File: trunk/tools/firefox-plugin/trump/chrome/content/uicreator/node.js

function NodeObject(){

    this.constants = {TAG : "tag"};
    this.id = null;
    this.xpath = null;
    this.attributes = new HashMap();
    this.parent = null;
    this.children = new Array();
    this.ui = new Ui();

}

NodeObject.prototype.getLevel = function(){
    var level = 0;
    var current = this;
    while(current.parent != null){
        level++;
        current = current.parent;
    }
}

NodeObject.prototype.printUI = function(){
    var hasChildren = false;
        if(this.children.length > 0)
            hasChildren = true;

        var sb = new StringBuffer();

        //get the current level of the node so that we can do pretty print
        var level = this.getLevel();

        for(var i=0; i<level; i++){
            sb.append("  ");
        }
        sb.append(this.ui.getType(this.attributes.get(this.constants.TAG), hasChildren)).append("(UID: '").append(this.id).append("', clocator: [");

        if(this.attributes.size() == 0){
            sb.append(":");
        }else{
            var count = 0;
            var attr;
            var keySet = this.attributes.keySet();
            for(var s=0; s < keySet.length; ++s){
                attr = this.attributes.get[keySet[s]];
                alert("keySet[s] : " + keySet[s]);
                alert("attr : " + attr);
                if(++count > 1){
                     sb.append(",");
                }
                sb.append(keySet[s]).append(":").append("'").append(attr).append("'");
            }
        }

        sb.append("]");

        //comment this line out if you do not want xpath to display
//        sb.append("[xpath: ").append(xpath).append("]");

        sb.append(")");

        if(hasChildren)
            sb.append("{");

        alert(sb.toString());

        if(hasChildren){
            for(var a=0; a<this.children.length; ++a){
                this.children[a].printUI();
            }

        }
        var indent = new StringBuffer();
        if(hasChildren){
            for(var i=0; i<level; i++){
                indent.append("  ");
            }
            indent.append("}");
            alert(indent.toString());
        }
    return sb.toString() + indent.toString();
}

NodeObject.prototype.isEmpty = function(){
    return (this.children != null && this.children.length > 0);
}

NodeObject.prototype.addChild = function(child){
    this.children.push(child);
}

NodeObject.prototype.removeChild = function(uid){
    var child = this.findChild(uid);
    if (child != null) {
        var index = this.children.indexOf(child);
        this.children.splice(index, 1);
    }
}

NodeObject.prototype.findChild = function(uid){
    var current;
    for(var i=0; i<this.children.length ; ++i){
        current = this.children[i];
        if(current.id == uid){
            return current;
        }
    }
    return null;
}